### ISSUE TITLE
No error output, but config.plist not being converted to config.plist.txt

### ISSUE THREAD

**Issue by snoby** [rusty-bits/OC-tool#21]
I thought I was really good at bash scripts, then started to dig into parsePlistEdit.sh.  OMG.
anyway I have the attached config.plist and the parsePlistEdit.sh script only does a partial generation of config.plist.txt.  Looking at the Errors.txt it's complaining about a section not being complete, but this is an AMD build and according to the docs this is a valid configuration.

attached is my config.plist renamed to config.plist.txt so that I can attach to this ticket.
[config.plist.txt](https://github.com/rusty-bits/OC-tool/files/4293081/config.plist.txt)

**Comment by rusty-bits** [comment]
You might be better than I am at bash scripts, this is my first attempt.  I wanted to learn shell scripting and thought this project would be a fun way to do it.

So, you're probably right about your config.plist being valid for an AMD build, but as I mentioned over in issue #20 some choices I made require certain sections of the plist to have data for OC-tool to work properly.
I went by https://github.com/acidanthera/OpenCorePkg/blob/master/Docs/Configuration.pdf 
particularly this part
![Screen Shot 2020-03-05 at 10 00 52 AM](https://user-images.githubusercontent.com/46965981/75996325-b833dc80-5ecb-11ea-96f1-8118568bdef6.png)

and let OC-tool run with the assumption that all sections would have a value, well we see how that is turning out

I have two ideas, and one other project, in the works. I can
1) rewrite parts of OC-tool to recognize and ignore missing sections
   I had something like that in there already and took it out for some reason, can't exactly remember why...
2) have OC-tool merge the missing sections into the temp/config.plist to avoid potential errors

I'm leaning towards number 2 right now which should get OC-tool working fine without causing any issues with booting, etc.

This will let me focus on my latest project, I want to learn the D-language, so I'm going to start a compiled version of OC-tool to help me learn that.

**Comment by snoby** [comment]
So if you merge the section... how will that not mess up the boot process?
[From this page]9https://khronokernel-2.gitbook.io/opencore-vanilla-desktop-guide/amd-config.plist/amd-config)
> Add: Sets device properties from a map.
By default, the Sample.plist has this section set for iGPU and Audio. We have no iGPU so PCIRoot PciRoot(0x0)/Pci(0x2,0x0) can be removed from Add section. For audio we'll be setting the layout in the boot-args section, so removal of PciRoot(0x0)/Pci(0x1b,0x0) is also recommended from both Add and Block sections
TL;DR, delete all the PciRoot's here as we won't be using this section.


Also side bar:  D language seems intense.  I'm partial to Golang, but that's mainly because I'm from 15 years of embedded C development and it's just how my brain works.

**Comment by snoby** [comment]
Also I merged those two device properties into the config and it still didn't generate the flattened config.plist.txt file.

```
cat errors.minor.txt
Line number 581
Found empty <array> in Misc    BlessOverride
This may not be an issue for certain sections

Line number 8
Found empty <array> in ACPI    Add
This may not be an issue for certain sections

Line number 10
Found empty <array> in ACPI    Block
This may not be an issue for certain sections

Line number 12
Found empty <array> in ACPI    Patch
This may not be an issue for certain sections

Line number 30
Found empty <array> in Booter    MmioWhitelist
This may not be an issue for certain sections

Line number 89
Found empty <array> in Kernel    Add
This may not be an issue for certain sections

Line number 91
Found empty <array> in Kernel    Block
This may not be an issue for certain sections

Line number 102
Found empty <array> in Kernel    Patch
This may not be an issue for certain sections

Line number 138
Found empty <array> in Misc    BlessOverride
This may not be an issue for certain sections

Line number 174
Found empty <array> in Misc    Entries
This may not be an issue for certain sections

Line number 193
Found empty <array> in Misc    Tools
This may not be an issue for certain sections
```


[tempfiles.zip](https://github.com/rusty-bits/OC-tool/files/4293972/tempfiles.zip)

**Comment by rusty-bits** [comment]
pushed commit cee6ba5 which should fix these issues

**Comment by rusty-bits** [comment]
Also, Golang??? it takes so loooooooooong to compile!  (not) 😉

**Comment by snoby** [comment]
Ok it's much better now.
```
Processing config.plist ... done
Checking for missing fields ... done
Checking for extra fields ... done
Using /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/INPUT/config.plist

Writing new config.plist to /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/INPUT/modified.config.plist ... done

Downloading/unzipping needed or updated resources
Downloading HFSPlus.efi ... unzipping ... done

Removing old /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/OUTPUT/EFI ... done
Creating new /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/OUTPUT/EFI ... done

Moving prebuiltRelease resources into place
Copying BOOTx64.efi to BOOT ... done
Copying OpenCore.efi to OC ... done
Copying config.plist to OC ... done
Copying Lilu.kext to OC/Kexts ... done
Copying VirtualSMC.kext to OC/Kexts ... done
Copying AppleMCEReporterDisabler.kext to OC/Kexts ... done
Copying SmallTreeIntel82576.kext to OC/Kexts ... done
Copying WhateverGreen.kext to OC/Kexts ... done
Copying HfsPlus.efi to OC/Drivers ... done
Copying ApfsDriverLoader.efi to OC/Drivers ... done
Copying FwRuntimeServices.efi to OC/Drivers ... done
Copied 11 resources to /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/OUTPUT/EFI

Finished building /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/OUTPUT/EFI

NOTE:	config.plist in /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/OUTPUT/EFI was changed by input scan
	this has been done to ensure OC wont fail on boot with this new EFI folder
	modified.config.plist has been saved in /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/INPUT
	the original /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/INPUT/config.plist has not been touched

NOTE:	issues were found in /Users/foo/Dropbox/home_work/osx-amd/round2/OC-tool/INPUT/config.plist
	Some may have been repaired by the input scan.
	Some may not be important. You can view the issues with
	'cat last_run_issues.txt'

~/Dropbox/home_work/osx-amd/round2/OC-tool[master]$ cat last_run_issues.txt
Line number 66
Found empty <dict> in DeviceProperties    Add
It is recommended that all sections be complete

Line number 68
Found empty <dict> in DeviceProperties    Block
It is recommended that all sections be complete

```

**Comment by rusty-bits** [comment]
nice, one missing `ds=""` can lead to a big mess, thanks for the report
